File: java/src/tarehart/rlbot/planning/Goal.java
package tarehart.rlbot.planning;

import tarehart.rlbot.math.Plane;
import tarehart.rlbot.math.Polygon;
import tarehart.rlbot.math.vector.Vector3;
import tarehart.rlbot.physics.ArenaModel;


public class Goal {

    private static final double GOAL_DISTANCE = 102;
    public static final double GOAL_HEIGHT = 12;
    public static final double EXTENT = 17.8555;

    private Vector3 center;
    private Plane threatPlane;
    private Plane scorePlane;
    private Polygon box;

    public Goal(boolean negativeSide) {

        center = new Vector3(0, GOAL_DISTANCE * (negativeSide ? -1 : 1), 0);

        threatPlane = new Plane(new Vector3(0, negativeSide ? 1 : -1, 0), new Vector3(0, (GOAL_DISTANCE - 1) * (negativeSide ? -1 : 1), 0));
        scorePlane = new Plane(new Vector3(0, negativeSide ? 1 : -1, 0), new Vector3(0, (GOAL_DISTANCE + 2) * (negativeSide ? -1 : 1), 0));
        box = negativeSide ? ZoneDefinitions.BLUEBOX : ZoneDefinitions.ORANGEBOX;
    }


    public Vector3 getNearestEntrance(Vector3 ballPosition, double padding) {

        double adjustedExtent = EXTENT - ArenaModel.BALL_RADIUS - padding;
        double adjustedHeight = GOAL_HEIGHT - ArenaModel.BALL_RADIUS - padding;
        double x = Math.min(adjustedExtent, Math.max(-adjustedExtent, ballPosition.getX()));
        double z = Math.min(adjustedHeight, Math.max(ArenaModel.BALL_RADIUS, ballPosition.getZ()));
        return new Vector3(x, center.getY(), z);
    }

    public Plane getThreatPlane() {
        return threatPlane;
    }

    public Plane getScorePlane() {
        return scorePlane;
    }

    public Vector3 getCenter() {
        return center;
    }

    /**
     * From shooter's perspective
     */
    public Vector3 getLeftPost() {
        return getLeftPost(0);
    }

    /**
     * From shooter's perspective
     */
    public Vector3 getLeftPost(double padding) {
        return new Vector3(center.getX() - (EXTENT - padding) * Math.signum(center.getY()), center.getY(), center.getZ());
    }

    /**
     * From shooter's perspective
     */
    public Vector3 getRightPost() {
        return getRightPost(0);
    }

    /**
     * From shooter's perspective
     */
    public Vector3 getRightPost(double padding) {
        return new Vector3(center.getX() + (EXTENT - padding) * Math.signum(center.getY()), center.getY(), center.getZ());
    }

    public boolean isInBox(Vector3 position) {
        return box.contains(position.flatten());
    }
}

File: src/main/java/tarehart/rlbot/planning/Goal.kt
package tarehart.rlbot.planning

import tarehart.rlbot.bots.Team
import tarehart.rlbot.math.BallSlice
import tarehart.rlbot.math.Plane
import tarehart.rlbot.math.vector.Vector3
import tarehart.rlbot.physics.BallPath


abstract class Goal( negativeSide: Boolean) {

    abstract val center: Vector3
    abstract val scorePlane: Plane
    val team: Team = if (negativeSide) Team.BLUE else Team.ORANGE

    /**
     * From shooter's perspective
     */
    val leftPost: Vector3
        get() = getLeftPost(0.0)

    /**
     * From shooter's perspective
     */
    val rightPost: Vector3
        get() = getRightPost(0.0)

    abstract fun getNearestEntrance(ballPosition: Vector3, padding: Double): Vector3

    /**
     * From shooter's perspective
     */
    abstract fun getLeftPost(padding: Double): Vector3

    /**
     * From shooter's perspective
     */
    abstract fun getRightPost(padding: Double): Vector3

    abstract fun predictGoalEvent(ballPath: BallPath): BallSlice?
}
